# Numerical Linear Algebra

## Solving Matrix Equations

Consider the system of linear equations
\begin{equation}
\begin{aligned}
a x + b y &= p, \\
c x + d y &= q.
\end{aligned}
\end{equation}

We can write this as the matrix equation
\begin{equation}
\mathbf{M}\, \vec{x} = \vec{r},
\end{equation}
or more explicitly,
\begin{equation}
\begin{bmatrix}
a & b \\ c & d \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} =
\begin{bmatrix} p \\ q \end{bmatrix}.
\end{equation}


## Elementary Row Operations

To solve matrix equations of the form shown above, we need to make use of the three elementary row operations:
<ol>
<li> interchange of two rows $R_i, R_j \to R_j, R_i$;
<li> scaling of a row $R_i \to \alpha R_i$;
<li> addition of two rows $R_i \to R_i + R_j$;
</ol>

These operations do not change the solution of the problem.

## Gaussian Elimination

Gaussian elimination make use of successive elementary row operations to solve a given matrix equation. The goal is to convert the matrix $\mathbf{M}$ into an identity matrix, or to a <em>reduced row echelon form</em>, or even just to <em>row echelon form</em>.

For example, the matrix
\begin{equation}
\mathbf{P} = \begin{bmatrix}
1 & p_{12} & p_{13} & \cdots & p_{1N} \\
0 & 1 & p_{23} & \cdots & p_{2N} \\
0 & 0 & 1 & \cdots & p_{3N} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & \cdots & 1 \end{bmatrix}
\end{equation}
is of reduced row echelon form, because the diagonals are all 1, the lower triangle of the matrix are all zeros, the non-zero matrix elements are in the upper triangle. A matrix that has only zero matrix elements in the lower triangle, but whose diagonals are not ones, is called to have a <em>row echelon form</em>.

To see how this works, consider the matrix equation $\mathbf{A}\vec{x} = \vec{b}$, which we write more explicitly as
\begin{equation}
\begin{bmatrix}
2 & -3 & 1 & 3 \\
1 & 4 & -3 & -3 \\
5 & 3 & -1 & -1 \\
3 & -6 & -3 & 1 \end{bmatrix}
\begin{bmatrix} w \\ x \\ y \\ z \end{bmatrix} =
\begin{bmatrix} -4 \\ 1 \\ 8 \\ -5 \end{bmatrix}.
\end{equation}

In [2]:
import numpy as np
A = np.array([[2.0,-3.0,1.0,3.0],[1.0,4.0,-3.0,-3.0],[5.0,3.0,-1.0,-1.0],[3.0,-6.0,-3.0,1.0]])
b = np.array([[-4.0],[1.0],[8.0],[-5.0]])
A, b

(array([[ 2., -3.,  1.,  3.],
        [ 1.,  4., -3., -3.],
        [ 5.,  3., -1., -1.],
        [ 3., -6., -3.,  1.]]), array([[-4.],
        [ 1.],
        [ 8.],
        [-5.]]))

First construct the <em>augmented matrix</em>,

In [2]:
Ab = np.concatenate((A, b), axis=1)2
Ab

array([[ 2., -3.,  1.,  3., -4.],
       [ 1.,  4., -3., -3.,  1.],
       [ 5.,  3., -1., -1.,  8.],
       [ 3., -6., -3.,  1., -5.]])

Now, let us do Gaussian elimination without pivoting to the identity matrix. In fact, we can do this without any interchange of rows.

First, let us rescale the first row, to make its leading element 1:

In [3]:
Ab[0,:] = Ab[0,:]/2.0
Ab

array([[ 1. , -1.5,  0.5,  1.5, -2. ],
       [ 1. ,  4. , -3. , -3. ,  1. ],
       [ 5. ,  3. , -1. , -1. ,  8. ],
       [ 3. , -6. , -3. ,  1. , -5. ]])

Next, we subtract an appropriate multiple of the first row from the subsequent rows:

In [4]:
Ab[1,:] = Ab[1,:] - Ab[0,:]
Ab[2,:] = Ab[2,:] - 5.0*Ab[0,:]
Ab[3,:] = Ab[3,:] - 3.0*Ab[0,:]
Ab

array([[  1. ,  -1.5,   0.5,   1.5,  -2. ],
       [  0. ,   5.5,  -3.5,  -4.5,   3. ],
       [  0. ,  10.5,  -3.5,  -8.5,  18. ],
       [  0. ,  -1.5,  -4.5,  -3.5,   1. ]])

Following this, we rescale the second row, so that its leading element is 1:

In [5]:
Ab[1,:] = Ab[1,:]/5.5
Ab

array([[  1.        ,  -1.5       ,   0.5       ,   1.5       ,  -2.        ],
       [  0.        ,   1.        ,  -0.63636364,  -0.81818182,
          0.54545455],
       [  0.        ,  10.5       ,  -3.5       ,  -8.5       ,  18.        ],
       [  0.        ,  -1.5       ,  -4.5       ,  -3.5       ,   1.        ]])

We then subtract appropriate multiples of row 2 from row 3 and row 4:

In [6]:
Ab[2,:] = Ab[2,:] - 10.5*Ab[1,:]
Ab[3,:] = Ab[3,:] + 1.5*Ab[1,:]
Ab

array([[  1.        ,  -1.5       ,   0.5       ,   1.5       ,  -2.        ],
       [  0.        ,   1.        ,  -0.63636364,  -0.81818182,
          0.54545455],
       [  0.        ,   0.        ,   3.18181818,   0.09090909,
         12.27272727],
       [  0.        ,   0.        ,  -5.45454545,  -4.72727273,
          1.81818182]])

Continuing, we rescale row 3 so that its leading element is 1:

In [7]:
Ab[2,:] = Ab[2,:]/Ab[2,2]
Ab

array([[ 1.        , -1.5       ,  0.5       ,  1.5       , -2.        ],
       [ 0.        ,  1.        , -0.63636364, -0.81818182,  0.54545455],
       [ 0.        ,  0.        ,  1.        ,  0.02857143,  3.85714286],
       [ 0.        ,  0.        , -5.45454545, -4.72727273,  1.81818182]])

And subtract an appropriate multiple of row 3 from row 4:

In [8]:
Ab[3,:] = Ab[3,:] - Ab[3,2]*Ab[2,:]
Ab

array([[  1.        ,  -1.5       ,   0.5       ,   1.5       ,  -2.        ],
       [  0.        ,   1.        ,  -0.63636364,  -0.81818182,
          0.54545455],
       [  0.        ,   0.        ,   1.        ,   0.02857143,
          3.85714286],
       [  0.        ,   0.        ,   0.        ,  -4.57142857,
         22.85714286]])

We then rescale row 4:

In [9]:
Ab[3,:] = Ab[3,:]/Ab[3,3]
Ab

array([[ 1.        , -1.5       ,  0.5       ,  1.5       , -2.        ],
       [ 0.        ,  1.        , -0.63636364, -0.81818182,  0.54545455],
       [ 0.        ,  0.        ,  1.        ,  0.02857143,  3.85714286],
       [-0.        , -0.        , -0.        ,  1.        , -5.        ]])

The augmented matrix is now in reduced row echelon form. If we wish, we can already solve for $[w, x, y, z]$ through backward substitution. But let us proceed to make the first part of the augmented matrix an identity matrix. This is done by subtracting an appropriate multiple of row 4 from row 3:

In [10]:
Ab[2,:] = Ab[2,:] - Ab[2,3]*Ab[3,:]
Ab

array([[ 1.        , -1.5       ,  0.5       ,  1.5       , -2.        ],
       [ 0.        ,  1.        , -0.63636364, -0.81818182,  0.54545455],
       [ 0.        ,  0.        ,  1.        ,  0.        ,  4.        ],
       [-0.        , -0.        , -0.        ,  1.        , -5.        ]])

followed by subtracting an appropriate multiple of row 4 from row 2, and an appropriate multiple of row 3 from row 2:

In [11]:
Ab[1,:] = Ab[1,:] - Ab[1,3]*Ab[3,:]
Ab[1,:] = Ab[1,:] - Ab[1,2]*Ab[2,:]
Ab

array([[ 1. , -1.5,  0.5,  1.5, -2. ],
       [ 0. ,  1. ,  0. ,  0. , -1. ],
       [ 0. ,  0. ,  1. ,  0. ,  4. ],
       [-0. , -0. , -0. ,  1. , -5. ]])

Finally, we subtract appropriate multiples of row 4, row 3, and row 2 from row 1:

In [12]:
Ab[0,:] = Ab[0,:] - Ab[0,3]*Ab[3,:]
Ab[0,:] = Ab[0,:] - Ab[0,2]*Ab[2,:]
Ab[0,:] = Ab[0,:] - Ab[0,1]*Ab[1,:]
Ab

array([[ 1.,  0.,  0.,  0.,  2.],
       [ 0.,  1.,  0.,  0., -1.],
       [ 0.,  0.,  1.,  0.,  4.],
       [-0., -0., -0.,  1., -5.]])

In [17]:
Ab[3,2]

-0.0

In this final form of the augmented matrix, we then read off the solution as $w = 2$, $x = -1$, $y = 4$, and $z = -5$.

## Gaussian Elimination with Pivoting

In the above example, we went through the Gaussian elimination process without any row interchanges. When the system of linear equations is large, involving $N \gg 1$ variables, or when the system of linear equations is <em>stiff</em>, it is generally advisable to perform pivoting during the Gaussian elimination.

Let us try the same example again, with pivoting.

In [18]:
Ab = np.concatenate((A, b), axis=1)
Ab

array([[ 2., -3.,  1.,  3., -4.],
       [ 1.,  4., -3., -3.,  1.],
       [ 5.,  3., -1., -1.,  8.],
       [ 3., -6., -3.,  1., -5.]])

To performing pivoting, we must ensure that when we are rescaling the leading matrix element, we rescale the row with the largest absolute matrix element. In this case, we should make $[5, 3, -1, -1, 8]$ the first row, and rescale it by 5:

In [19]:
Ab[[0,2]] = Ab[[2,0]]
Ab

array([[ 5.,  3., -1., -1.,  8.],
       [ 1.,  4., -3., -3.,  1.],
       [ 2., -3.,  1.,  3., -4.],
       [ 3., -6., -3.,  1., -5.]])

In [20]:
Ab[0,:] = Ab[0,:]/Ab[0,0]
Ab

array([[ 1. ,  0.6, -0.2, -0.2,  1.6],
       [ 1. ,  4. , -3. , -3. ,  1. ],
       [ 2. , -3. ,  1. ,  3. , -4. ],
       [ 3. , -6. , -3. ,  1. , -5. ]])

Once this is done, we subtract appropriate multiples of row 1 from rows 2, 3, 4:

In [21]:
Ab[1,:] = Ab[1,:] - Ab[1,0]*Ab[0,:]
Ab[2,:] = Ab[2,:] - Ab[2,0]*Ab[0,:]
Ab[3,:] = Ab[3,:] - Ab[3,0]*Ab[0,:]
Ab

array([[ 1. ,  0.6, -0.2, -0.2,  1.6],
       [ 0. ,  3.4, -2.8, -2.8, -0.6],
       [ 0. , -4.2,  1.4,  3.4, -7.2],
       [ 0. , -7.8, -2.4,  1.6, -9.8]])

Next, we rescale the leading element of the second row. To pivot, we must first interchange row 2 and row 4, since $|-7.8|$ is the largest in magnitude:

In [22]:
Ab[[1,3]] = Ab[[3,1]]
Ab

array([[ 1. ,  0.6, -0.2, -0.2,  1.6],
       [ 0. , -7.8, -2.4,  1.6, -9.8],
       [ 0. , -4.2,  1.4,  3.4, -7.2],
       [ 0. ,  3.4, -2.8, -2.8, -0.6]])

In [23]:
Ab[1,:] = Ab[1,:]/Ab[1,1]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [ 0.        , -4.2       ,  1.4       ,  3.4       , -7.2       ],
       [ 0.        ,  3.4       , -2.8       , -2.8       , -0.6       ]])

To proceed, we subtract appropriate multiples of row 2 from rows 3 and 4:

In [24]:
Ab[2,:] = Ab[2,:] - Ab[2,1]*Ab[1,:]
Ab[3,:] = Ab[3,:] - Ab[3,1]*Ab[1,:]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [ 0.        ,  0.        ,  2.69230769,  2.53846154, -1.92307692],
       [ 0.        ,  0.        , -3.84615385, -2.1025641 , -4.87179487]])

For the next rescaling, we interchange row 3 and row 4:

In [25]:
Ab[[2,3]] = Ab[[3,2]]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [ 0.        ,  0.        , -3.84615385, -2.1025641 , -4.87179487],
       [ 0.        ,  0.        ,  2.69230769,  2.53846154, -1.92307692]])

In [26]:
Ab[2,:] = Ab[2,:]/Ab[2,2]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [-0.        , -0.        ,  1.        ,  0.54666667,  1.26666667],
       [ 0.        ,  0.        ,  2.69230769,  2.53846154, -1.92307692]])

We are almost there! Subtracting an appropriate multiple of row 3 from row 4:

In [27]:
Ab[3,:] = Ab[3,:] - Ab[3,2]*Ab[2,:]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [-0.        , -0.        ,  1.        ,  0.54666667,  1.26666667],
       [ 0.        ,  0.        ,  0.        ,  1.06666667, -5.33333333]])

Finally, we rescale row 4:

In [28]:
Ab[3,:] = Ab[3,:]/Ab[3,3]
Ab

array([[ 1.        ,  0.6       , -0.2       , -0.2       ,  1.6       ],
       [-0.        ,  1.        ,  0.30769231, -0.20512821,  1.25641026],
       [-0.        , -0.        ,  1.        ,  0.54666667,  1.26666667],
       [ 0.        ,  0.        ,  0.        ,  1.        , -5.        ]])

and we end up with a (different) reduced row echelon form.

From this augmented matrix, we read off the solution for $z$ from the last row as $z = -5$.

The third row of this augmented matrix tells us that $y + 0.546666667 z = 1.26666667$. Therefore, to solve for $y$,

In [29]:
Ab[2,4] - Ab[2,3]*Ab[3,4]

4.0000000000000009

which is indeed what we get from the previous Gaussian elimination.

Moving on, the second row of the augmented matrix tells us that
\begin{equation}
x + 0.30769231 y - 0.20512821 z = 1.25641026,
\end{equation}
and therefore $x$ is:

In [30]:
Ab[1,4] - Ab[1,2]*4 - Ab[1,3]*(-5)

-1.0

Without writing out the equation for $w$, we can see that its solution must be:

In [31]:
Ab[0,4] - Ab[0,1]*(-1.0) - Ab[0,2]*(4.0) - Ab[0,3]*(-5.0)

2.0

## Solving a Matrix Equation Using the Numpy Solver

To ensure that you do not become button-pushing technicians, I have shown you the basic principles behind Gaussian elimination in solving matrix equations.

Naturally, for implementing implicit schemes to solve PDEs, you should not need to do the Gaussian elimination manually as shown above, or even write your own Gaussian elimination solver.

Instead, you should use the built-in solver in numpy, as shown below:

In [43]:
np.linalg.solve(A,b)

array([[ 2.],
       [-1.],
       [ 4.],
       [-5.]])

## Other Methods for Solving Matrix Equations

### Cramers Method

Most of you would have learnt Cramers method for solving matrix equations. In the Cramers method, you compute <em>determinants</em> instead of performing Gaussian elimination.

For a $N \times N$ matrix
\begin{equation}
\mathbf{A} = \begin{bmatrix}
A_{11} & A_{12} & A_{13} & \cdots & A_{1N} \\
A_{21} & A_{22} & A_{23} & \cdots & A_{2N} \\
A_{31} & A_{32} & A_{33} & \cdots & A_{3N} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
A_{N1} & A_{N2} & A_{N3} & \cdots & A_{NN}
\end{bmatrix},
\end{equation}
its determinant is defined as
\begin{equation}
\det\mathbf{A} = |\mathbf{A}| = \sum_{\pi} (-1)^{\pi} \prod_{i=1}^N A_{i,\pi(i)},
\end{equation}
where the sum is over all permutations $\pi$ of $(1, 2, \dots, N)$, and
\begin{equation}
(-1)^{\pi} = \begin{cases}
1, & \text{$\pi$ is an even permutation}; \\
-1, & \text{$\pi$ is an odd permutation}. \end{cases}
\end{equation}

For example, $(2, 3, 4, 1)$ is an odd permutation of $(1, 2, 3, 4)$, because
\begin{equation}
(2, 3, 4, 1) \to (2, 3, 1, 4) \to (2, 1, 3, 4) \to (1, 2, 3, 4).
\end{equation}

On the other hand, $(3, 1, 2, 4)$ is an even permutation of $(1, 2, 3, 4)$ because
\begin{equation}
(3, 1, 2, 4) \to (1, 3, 2, 4) \to (1, 2, 3, 4).
\end{equation}

Again, in a Computational Physics class, I do not expect you to evaluate the determinant of a matrix starting from the definition, but to use the built-in numpy function.

In [32]:
np.linalg.det(A)

-159.99999999999994

To apply the Cramers method to the matrix equation $\mathbf{A}\vec{x} = \vec{b}$, we need to first evaluate five determinants.

The first is the determinant of $\mathbf{A}$:

In [43]:
dA = np.linalg.det(A)
dA

-159.99999999999994

The next determinant we need to evaluate is when we replace the first column of $\mathbf{A}$ by $\vec{b}$:

In [33]:
Aw = np.concatenate((b,A[:,1:4]),axis=1)
Aw

array([[-4., -3.,  1.,  3.],
       [ 1.,  4., -3., -3.],
       [ 8.,  3., -1., -1.],
       [-5., -6., -3.,  1.]])

In [34]:
dAw = np.linalg.det(Aw)
dAw

-319.99999999999994

The third determinant we need to evaluate is when we replace the second column of $\mathbf{A}$ by $\vec{b}$:

In [35]:
Ax = np.array(A)
Ax

array([[ 2., -3.,  1.,  3.],
       [ 1.,  4., -3., -3.],
       [ 5.,  3., -1., -1.],
       [ 3., -6., -3.,  1.]])

In [36]:
Ax[:,1:2] = b
Ax

array([[ 2., -4.,  1.,  3.],
       [ 1.,  1., -3., -3.],
       [ 5.,  8., -1., -1.],
       [ 3., -5., -3.,  1.]])

In [37]:
dAx = np.linalg.det(Ax)
dAx

159.99999999999994

In [38]:
Ay = np.array(A)
Ay[:,2:3] = b
Ay

array([[ 2., -3., -4.,  3.],
       [ 1.,  4.,  1., -3.],
       [ 5.,  3.,  8., -1.],
       [ 3., -6., -5.,  1.]])

In [39]:
dAy = np.linalg.det(Ay)
dAy

-639.99999999999989

In [40]:
Az = np.array(A)
Az[:,3:4] = b
Az

array([[ 2., -3.,  1., -4.],
       [ 1.,  4., -3.,  1.],
       [ 5.,  3., -1.,  8.],
       [ 3., -6., -3., -5.]])

In [41]:
dAz = np.linalg.det(Az)
dAz

799.99999999999989

Finally, Cramers method tells us that the solution is:

In [44]:
w = dAw/dA
w

2.0000000000000004

In [45]:
x = dAx/dA
x

-1.0

In [46]:
y = dAy/dA
y

4.0000000000000009

In [47]:
z = dAz/dA
z

-5.0000000000000009

<b>Remark:</b> The Cramers method is not suitable for solving large matrix equations, because of the large number of determinants that must first be evaluated.

### LU Decomposition

If we have to solve the matrix equation $\mathbf{A}\vec{x} = \vec{b}$ once, then Gaussian elimination would be fine.

However, if we have to keep solving the matrix equation $\mathbf{A}\vec{x} = \vec{b}_i$, for $i = 1, 2, \dots, M$, then it would be better to first perform an <em>LU Decomposition</em>.


In a LU Decomposition, we write the $N \times N$ matrix
\begin{equation}
\mathbf{A} = \mathbf{L}\mathbf{U}
\end{equation}
as the product of two $N \times N$ matrices $\mathbf{L}$ and $\mathbf{U}$.

The matrix
\begin{equation}
\mathbf{L} = \begin{bmatrix}
L_{11} & 0 & 0 & \cdots & 0 \\
L_{21} & L_{22} & 0 & \cdots & 0 \\
L_{31} & L_{32} & L_{33} & \cdots & 0 \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
L_{N1} & L_{N2} & L_{N3} & \cdots & L_{NN} \end{bmatrix}
\end{equation}
is a <em>lower triangular</em> matrix, whereas the matrix
\begin{equation}
\mathbf{U} = \begin{bmatrix}
U_{11} & U_{12} & U_{13} & \cdots & U_{1N} \\
0 & U_{22} & U_{23} & \cdots & U_{2N} \\
0 & 0 & U_{33} & \cdots & U_{3N} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & \cdots & U_{NN} \end{bmatrix}
\end{equation}
is an <em>upper triangular</em> matrix.

In terms of $\mathbf{L}$ and $\mathbf{U}$, the original matrix equation can be written as
\begin{equation}
\mathbf{A}\vec{x} = \mathbf{L}\mathbf{U}\vec{x} = \mathbf{L}\vec{y} = \vec{b}.
\end{equation}
This tells us that we can solve for $\vec{y} = (y_1, y_2, \dots, y_N)$ from $\mathbf{L}\vec{y} = \vec{b}$ using forward substutition.

Once this is done, we can solve for $\vec{x} = (x_1, x_2, \dots, x_N)$ from $\mathbf{U}\vec{x} = \vec{y}$ using backward substitution.

### Cholesky Decomposition

If $\mathbf{A}$ is symmetric, i.e. $\mathbf{A}^T = \mathbf{A}$, then 
\begin{equation}
\mathbf{A} = \mathbf{L}\mathbf{L}^T,
\end{equation}
making the decomposition of $\mathbf{A}$ into a lower triangular matrix $\mathbf{L}$ and an upper triangular matrix $\mathbf{L}^T$ easier.

This decomposition is then known as <em>Cholesky decomposition</em>.

# Eigenvalue Problem

The other matrix problem we commonly encounter in computational physics is the <em>eigenvalue problem</em>,
\begin{equation}
\mathbf{A}\vec{x} = \lambda\vec{x},
\end{equation}
where $\mathbf{A}$ is a given matrix, and we are to solve for possible pairs of $\vec{x}$ (the <em>eigenvector</em>) and $\lambda$ (the <em>eigenvalue</em>).

In general, if $\mathbf{A}$ is $N \times N$, we can find $N$ eigenvectors $\vec{x}_i$ and their $N$ eigenvalues $\lambda_i$. Sometimes, it is also possible to find fewer than $N$ eigenvectors.

Let us use
\begin{equation}
\mathbf{A} = \begin{bmatrix}
2 & -3 & 1 & 3 \\
1 & 4 & -3 & -3 \\
5 & 3 & -1 & -1 \\
3 & -6 & -3 & 1 \end{bmatrix}
\end{equation}
as an example, to learn how to find its eigenvectors and eigenvalues.

In general, the $N$ eigenvectors $\vec{u}_i$ of $\mathbf{A}$ are linearly independent, so we can write an arbitrary vector $\vec{x}$ in terms of them as
\begin{equation}
\vec{x} = \sum_{i=1}^N \alpha_i \vec{u}_i.
\end{equation}

Therefore, if we let $\mathbf{A}$ act on $\vec{x}$,
\begin{equation}
\mathbf{A}\vec{x} = \sum_{i=1}^N \alpha_i \mathbf{A}\vec{u}_i = \sum_{i=1}^N \alpha_i \lambda_i \vec{u}_i.
\end{equation}
If we multiply $\mathbf{A}\vec{x}$ by $\mathbf{A}$ again, we will have
\begin{equation}
\mathbf{A}(\mathbf{A}\vec{x}) = \mathbf{A}^2\vec{x} = \sum_{i=1}^N \alpha_i \lambda_i^2 \vec{u}_i.
\end{equation}
We see that, if we multiply $\vec{x}$ by $\mathbf{A}$ $M$ times, we will end up with
\begin{equation}
\mathbf{A}^M\vec{x} = \sum_{i=1}^N \alpha_i \lambda_i^M \vec{u}_i.
\end{equation}

## The Power Method

In general, $\mathbf{A}$ will have an eigenvalue $\lambda_1$ with the largest absolute value $|\lambda_1| > |\lambda_i|$, $i = 2, \dots, N$. If we pull this term out of the sum, we will have
\begin{equation}
\mathbf{A}^M\vec{x} = \alpha_1 \lambda_1^M \vec{u}_1 + \sum_{i=2}^N \alpha_i \lambda_i^M \vec{u}_i,
\end{equation}
which we can rewrite as
\begin{equation}
\mathbf{A}^M\vec{x} = \lambda_1^M \left[\alpha_1 \vec{u}_1 + \sum_{i=2}^N \alpha_i \left(\dfrac{\lambda_i^M}{\lambda_1^M}\right) \vec{u}_i\right].
\end{equation}

Now, $|\lambda_i/\lambda_1| < 1$ for all $i = 2, \dots, N$, so if $M$ is large enough, we will have
\begin{equation}
\lim_{M \to \infty} \mathbf{A}^M\vec{x} = \lim_{M \to \infty} \lambda_1^M \alpha_1 \vec{u}_1.
\end{equation}
In other words, whatever $\vec{x}$ we start with, after multiplying by $\mathbf{A}$ a large number of times, we would end up with a vector proportional to $\vec{u}_1$, the eigenvector of $\mathbf{A}$ associated with the eigenvalue $\lambda_1$ with the largest absolute value.

This suggests that to find $\vec{u}_1$ and $\lambda_1$, we start with a random $\vec{x}_0$, and calculate iteratively
\begin{equation}
\vec{x}_{i+1} = \dfrac{\mathbf{A}\vec{x}_i}{|\mathbf{A}\vec{x}_i|}.
\end{equation}
Then $\vec{u}_1 = \lim_{i \to \infty} \vec{x}_i$, and since $\vec{u}_1$ will be properly normalized, we find $\lambda_1$ by
\begin{equation}
\lambda_1 = \vec{u}_1^T \mathbf{A} \vec{u}_1.
\end{equation}

This method, which was invented during the Second World War, of finding the the eigenvector $\vec{u}_1$ associated with the eigenvalue $\lambda_1$ with the largest absolute value is called the <em>power method</em>.

In the following, let us test the power method on the matrix $\mathbf{A}$.

In [48]:
A

array([[ 2., -3.,  1.,  3.],
       [ 1.,  4., -3., -3.],
       [ 5.,  3., -1., -1.],
       [ 3., -6., -3.,  1.]])

In [4]:
x0 = np.random.random((4,1))

In [5]:
x0

array([[ 0.38182897],
       [ 0.92796689],
       [ 0.27018296],
       [ 0.60771336]])

In [6]:
x0 = x0/np.linalg.norm(x0)
x0

array([[ 0.3171759 ],
       [ 0.77083918],
       [ 0.22443431],
       [ 0.50481248]])

In [7]:
x1 = np.matmul(A,x0)
x1 = x1/np.linalg.norm(x1)
x1

array([[ 0.01184206],
       [ 0.23658211],
       [ 0.61821326],
       [-0.74946717]])

In [8]:
x2 = np.matmul(A, x1)
x2 = x2/np.linalg.norm(x2)
x2

array([[-0.47371299],
       [ 0.27649335],
       [ 0.18410851],
       [-0.8156295 ]])

In [9]:
x3 = np.matmul(A, x2)
x3 = x3/np.linalg.norm(x3)
x3

array([[-0.61381811],
       [ 0.38394319],
       [-0.1379016 ],
       [-0.67586841]])

In [10]:
x4 = np.matmul(A, x3)
x4 = x4/np.linalg.norm(x4)
x4

array([[-0.62660906],
       [ 0.46368863],
       [-0.15213677],
       [-0.60762517]])

In [11]:
x5 = np.matmul(A, x4)
x5 = x5/np.linalg.norm(x5)
x5

array([[-0.60774235],
       [ 0.46145871],
       [-0.12922643],
       [-0.63325005]])

In [12]:
x6 = np.matmul(A, x5)
x6 = x6/np.linalg.norm(x6)
x6

array([[-0.60751996],
       [ 0.46271336],
       [-0.11705363],
       [-0.63491282]])

In [13]:
for n in range(20):
    x6 = np.matmul(A, x6)
    x6 = x6/np.linalg.norm(x6)
    print(n, x6)
    

0 [[-0.60582699]
 [ 0.45836594]
 [-0.11756301]
 [-0.63957272]]
1 [[-0.60710237]
 [ 0.45949876]
 [-0.11778219]
 [-0.63750683]]
2 [[-0.60673667]
 [ 0.45892508]
 [-0.11834526]
 [-0.63816361]]
3 [[-0.60697573]
 [ 0.45932507]
 [-0.11821555]
 [-0.63767236]]
4 [[-0.60684127]
 [ 0.45915001]
 [-0.11826835]
 [-0.63791656]]
5 [[-0.60690234]
 [ 0.45923978]
 [-0.11822217]
 [-0.63780239]]
6 [[-0.6068695 ]
 [ 0.45919063]
 [-0.11824148]
 [-0.63786545]]
7 [[-0.60688673]
 [ 0.45921507]
 [-0.11823195]
 [-0.63783322]]
8 [[-0.60687809]
 [ 0.4592025 ]
 [-0.11823739]
 [-0.63784949]]
9 [[-0.60688258]
 [ 0.45920905]
 [-0.11823471]
 [-0.637841  ]]
10 [[-0.60688026]
 [ 0.4592057 ]
 [-0.11823608]
 [-0.63784536]]
11 [[-0.60688145]
 [ 0.45920742]
 [-0.11823536]
 [-0.63784312]]
12 [[-0.60688083]
 [ 0.45920654]
 [-0.11823573]
 [-0.63784428]]
13 [[-0.60688115]
 [ 0.45920699]
 [-0.11823554]
 [-0.63784368]]
14 [[-0.60688099]
 [ 0.45920676]
 [-0.11823564]
 [-0.63784399]]
15 [[-0.60688107]
 [ 0.45920688]
 [-0.11823559]
 [

After more than 20 multiplications of the random initial vector $\vec{x}_0$ by $\mathbf{A}$, we end up with $\vec{u}_1 = [-0.606881, 0.459207, -0.118236, -0.637844]^T$, which is correct up to at least the sixth decimal place.

Next, we compute the eigenvalue $\lambda_1$ as

In [14]:
np.matmul(np.transpose(x6), (np.matmul(A, x6)))

array([[ 7.61788479]])

## The Inverse Power Method

As we have seen, the power method allows us to find the eigenvector $\vec{u}_1$ of the matrix $\mathbf{A}$ associated with the eigenvalue $\lambda_1$ with the largest absolute value.

But what if we are not interested in this eigenvector?

What if we are interested in the eigenvector associated with the eigenvalue with the smallest absolute value?

To modify the power method to find this, let us first observe that $\mathbf{A}^{-1}$ shares the same eigenvectors $\{\vec{u}_i\}_{i=1}^N$ as $\mathbf{A}$. To show this, let us start from
\begin{equation}
\mathbf{A}\vec{u}_i = \lambda_i \vec{u}_i,
\end{equation}
where $\vec{u}_i$ is the eigenvector of $\mathbf{A}$ with eigenvalue $\lambda_i$. Next, we multiply both sides of the equation by $\mathbf{A}^{-1}$:
\begin{equation}
\mathbf{A}^{-1}\mathbf{A}\vec{u}_i = \lambda_i \mathbf{A}^{-1}\vec{u}_i.
\end{equation}
Since $\mathbf{A}^{-1}\mathbf{A} = \mathbf{I}$, we therefore have
\begin{equation}
\vec{u}_i = \lambda_i \mathbf{A}^{-1}\vec{u}_i.
\end{equation}
Dividing throughout by $\lambda_i$, and making $\mathbf{A}^{-1}\vec{u}_i$ the subject, we finally have
\begin{equation}
\mathbf{A}^{-1}\vec{u}_i = \dfrac{1}{\lambda_i}\vec{u}_i,
\end{equation}
which tells us that $\vec{u}_i$ is an eigenvector of $\mathbf{A}^{-1}$ with eigenvalue $\lambda_i^{-1}$.

Now, let us pretend that we have worked out $\mathbf{A}^{-1}$, and write the $(i+1)$th iterate $\vec{x}_{i+1}$ in terms of the $i$th iterate $\vec{x}_i$ as
\begin{equation}
\vec{x}_{i+1} = \mathbf{A}^{-1}\vec{x}_i,
\end{equation}
then we know that after multiplying $\vec{x}_0$ by $\mathbf{A}^{-1}$ $M$ times, we have
\begin{equation}
\left(\mathbf{A}^{-1}\right)^M \vec{x}_0 = \sum_{i=1}^N \dfrac{\alpha_i}{\lambda_i^M} \vec{u}_i.
\end{equation}

Instead of converging to the eigenvector with the eigenvalue $\lambda_i$ with the largest absolute magnitude, we find the power method using $\mathbf{A}^{-1}$ converging to the eigenvector associated with the eigenvalue $\lambda_i$ with the smallest absolute magnitude.

In practice, given $\mathbf{A}$ we do not proceed to evaluate $\mathbf{A}^{-1}$ (which can be done using Gaussian elimination). Instead, we write the iteration equation as
\begin{equation}
\mathbf{A}\vec{x}_{i+1} = \vec{x}_i.
\end{equation}
This means that instead of solving for $\vec{x}_{i+1}$ from $\vec{x}_i$ using matrix-vector multiplication, we solve for $\vec{x}_{i+1}$ from $\vec{x}_i$ using Gaussian elimination.

Nevertheless, because this method is essentially the power method applied to the inverse of $\mathbf{A}$, it is called the <em>inverse power method</em>.

[Python code]

In [18]:
# import numpy
import numpy as np
# initialize the matrix A
A = np.array([[ 2., -3.,  1.,  3.],
       [ 1.,  4., -3., -3.],
       [ 5.,  3., -1., -1.],
       [ 3., -6., -3.,  1.]])
# choose an initial random vector x0
x0 = np.random.random((4,1))
# normalize x0
x0 = x0/np.linalg.norm(x0)
# start iteration
x = x0
for n in range(100):
    x = np.linalg.solve(A,x)
    x = x/np.linalg.norm(x)
    if n > 90:
        print(n, x)


91 [[-0.24296159]
 [ 0.12422429]
 [-0.74905811]
 [ 0.60369689]]
92 [[-0.56820305]
 [-0.30076265]
 [-0.76588571]
 [ 0.01030583]]
93 [[-0.11508994]
 [-0.402638  ]
 [ 0.4146511 ]
 [-0.80789938]]
94 [[ 0.24520029]
 [-0.12208006]
 [ 0.75047624]
 [-0.60146379]]
95 [[  5.70548942e-01]
 [  3.06204104e-01]
 [  7.62045141e-01]
 [ -3.93637761e-04]]
96 [[ 0.11072428]
 [ 0.39996211]
 [-0.41992692]
 [ 0.80711327]]
97 [[-0.24743955]
 [ 0.11992934]
 [-0.75188481]
 [ 0.59921603]]
98 [[-0.57283904]
 [-0.31166974]
 [-0.7580398 ]
 [-0.00964689]]
99 [[-0.10640271]
 [-0.39729539]
 [ 0.42511971]
 [-0.80629279]]


## Shifted Inverse Power Method

In the power method which targets the eigenvalue with the largest absolute value, there is no easy way to find other eigenvectors and eigenvalues. In the inverse power method, however, this is now possible.

To see how this works, first observe that the matrix $\mathbf{A} - \alpha \mathbf{I}$ has eigenvalues $\lambda_i - \alpha$. This matrix has the same eigenvectors $\vec{u}_i$ as $\mathbf{A}$. Let us suppose that for the matrix $\mathbf{A}$, $\lambda_1$ is the eigenvalue with the smallest absolute value, while $\lambda_2$ is the eigenvalue with the second smallest absolute value.

Then, with a suitable choice of $\alpha$, it is possible to make $|\lambda_2 - \alpha| < |\lambda_1 - \alpha|$.

When this condition is satisfied, the inverse power method will find $\vec{u}_2$, the eigenvector associated with $\lambda_2 - \alpha$. Because of the shift by $\alpha$, the method is called the <em>shifted inverse power method</em>.

[Python code]

## QR Method

While discussing how we can solve matrix equations, I introduced the LU decomposition, where any matrix $\mathbf{A} = \mathbf{L}\mathbf{U}$ can be written as the product of a lower triangular matrix $\mathbf{L}$ and an upper triangular matrix $\mathbf{U}$.

It turns out that it is also possible to write any matrix $\mathbf{A}$ as the product
\begin{equation}
\mathbf{A} = \mathbf{Q}\mathbf{R}
\end{equation}
of an orthogonal matrix $\mathbf{Q}$, and an upper triangular matrix $\mathbf{R}$. In fact, one might say that this <em>QR decomposition</em> of the matrix $\mathbf{A}$ is more 'natural' than the LU decomposition, because it can be accomplished by applying a sequence of rotations to $\mathbf{A}$. This is also why $\mathbf{Q}$ is an orthogonal matrix.

In case you forgot, an orthogonal matrix has the property
\begin{equation}
\mathbf{Q}\mathbf{Q}^T = \mathbf{I} = \mathbf{Q}^T \mathbf{Q}.
\end{equation}

We are now ready to discuss the QR method for diagonalizing the matrix $\mathbf{A}$. This method is voted as one of the top ten algorithms of the 20th century, but it is rather unintuitive.

First, let us call $\mathbf{A}_0 = \mathbf{A}$ as our initial matrix. This matrix has the QR decomposition
\begin{equation}
\mathbf{A}_0 = \mathbf{Q}_1 \mathbf{R}_1.
\end{equation}

Next, define
\begin{equation}
\mathbf{A}_1 = \mathbf{R}_1 \mathbf{Q}_1.
\end{equation}
We call $\mathbf{A}_1$ the <em>QR transform</em> of $\mathbf{A}_0$.

From $\mathbf{A}_0 = \mathbf{Q}_1 \mathbf{R}_1$ we can write
\begin{equation}
\mathbf{Q}_1^T \mathbf{A}_0 = \mathbf{Q}_1^T \mathbf{Q}_1 \mathbf{R}_1 = \mathbf{R}_1.
\end{equation}
Substituting this into the expression for $\mathbf{A}_1$, we find that
\begin{equation}
\mathbf{A}_1 = \mathbf{R}_1 \mathbf{Q}_1 = \mathbf{Q}_1^T \mathbf{A}_0 \mathbf{Q}_1.
\end{equation}
This tells us that $\mathbf{A}_0$ and $\mathbf{A}_1$ are related to each other through a <em>similarity transform</em>. Mathematicians tell us that if $\mathbf{A}$ and $\mathbf{B}$ are related to each other through a <em>similarity transform</em>, the two matrices have the same set of eigenvalues.

In general, let us say that we have computed $\mathbf{A}_k$ and is ready to compute $\mathbf{A}_{k+1}$. We can find $\mathbf{A}_{k+1}$ from
\begin{equation}
\mathbf{A}_{k+1} = \mathbf{Q}_k^T \mathbf{A}_k \mathbf{Q}_k,
\end{equation}
where $\mathbf{A}_k = \mathbf{Q}_k \mathbf{R}_k$ and $\mathbf{A}_{k+1} = \mathbf{R}_k \mathbf{Q}_k$. We can expand this expression as
\begin{equation}
\mathbf{A}_{k+1} = \mathbf{Q}_k^T \mathbf{A}_k \mathbf{Q}_k = \mathbf{Q}_k^T \mathbf{Q}_{k-1}^T \mathbf{A}_{k-1} \mathbf{Q}_{k-1} \mathbf{Q}_k = \mathbf{Q}_k^T \mathbf{Q}_{k-1}^T \cdots \mathbf{Q}_1^T\mathbf{A}_0 \mathbf{Q}_1 \cdots \mathbf{Q}_{k-1} \mathbf{Q}_k.
\end{equation}

In general, if $\mathbf{Q}_1$ and $\mathbf{Q}_2$ are orthogonal matrices, then the product $\mathbf{Q}_1 \mathbf{Q}_2$ is also an orthogonal matrix. Therefore, we can write
\begin{equation}
\mathbf{A}_{k+1} = \mathbf{O}_k^T \mathbf{A}_0 \mathbf{O}_k,
\end{equation}
where
\begin{equation}
\mathbf{O}_k = \mathbf{Q}_1 \mathbf{Q}_2 \cdots \mathbf{Q}_k.
\end{equation}

Therefore, if we iterate infinitely many times, we would have
\begin{equation}
\mathbf{A}_{\infty} = \mathbf{O}_{\infty}^T \mathbf{A}_0 \mathbf{O}_{\infty}
\end{equation}

Here comes the unintuitive bit: mathematicians can prove that $\mathbf{A}_{\infty}$ is an upper triangular matrix!

QUESTION: What are the eigenvalues of an upper triangular matrix?

[Answer here]

The immense advantage the QR method has over the power method or the inverse power method is that we get all the eigenvalues in one fell swoop!

In fact, we also get all the eigenvectors in one fell swoop, after the iteration converges!

[Python exercise]

In [20]:
At = np.array(A)
for n in range(50):
    # QR decomposition of At
    Q, R = np.linalg.qr(At)
    # QR transform of At
    At = np.matmul(R, Q)
A, At

(array([[ 2., -3.,  1.,  3.],
        [ 1.,  4., -3., -3.],
        [ 5.,  3., -1., -1.],
        [ 3., -6., -3.,  1.]]),
 array([[  7.61788478e+00,   2.50063338e+00,  -1.30705666e+00,
           3.04767343e+00],
        [ -1.66137401e-13,  -3.92116067e+00,  -4.29500371e+00,
           2.03764831e-01],
        [ -4.59415084e-25,  -2.39888640e-11,  -1.53915921e+00,
           3.57571306e+00],
        [ -1.34483550e-25,  -3.79867555e-12,  -3.15195706e+00,
           3.84243510e+00]]))

## Subspace Methods for Diagonalization

Sometimes your matrix is so large that it takes a long time to completely diagonalize. Moreover, you only need a small number of eigenvectors and eigenvalues.

When this is so, use subspace methods like the <em>Lanczos method</em> or the <em>Arnoldi</em> method.

## Matrix Diagonalization in Numpy

As always, we learn the working principles behind various methods of matrix diagonalization so that we can hold our heads up high and call ourselves computational physicists.

But in practice, we will not write functions that do these diagonalization, especially since numpy has built-in functions.

[Python: use eig() to diagonalize A]

In [21]:
A

array([[ 2., -3.,  1.,  3.],
       [ 1.,  4., -3., -3.],
       [ 5.,  3., -1., -1.],
       [ 3., -6., -3.,  1.]])

In [23]:
l, u = np.linalg.eig(A)

In [24]:
l

array([-3.92116067+0.j        ,  1.15163794+2.00751206j,
        1.15163794-2.00751206j,  7.61788478+0.j        ])

In [25]:
u

array([[-0.24040213+0.j        , -0.23920438-0.18985937j,
        -0.23920438+0.18985937j,  0.60688104+0.j        ],
       [ 0.43805314+0.j        ,  0.08358075-0.25397695j,
         0.08358075+0.25397695j, -0.45920684+0.j        ],
       [ 0.24593406+0.j        , -0.67290854+0.j        ,
        -0.67290854-0.j        ,  0.11823560+0.j        ],
       [ 0.83056167+0.j        ,  0.50257590-0.36035568j,
         0.50257590+0.36035568j,  0.63784389+0.j        ]])